README.md
# SAM - System Architecture Mapper

SAM is a tool designed to map a network based on the data log of a router.
It runs as a local python-based server and displays the a map and statistics on the browser.

Check out the [website](http://sam.centralus.cloudapp.azure.com) for details about the project and a demo!

## Prerequisites

MySQL - SAM will support other databases in the future:

    apt-get install mysql-server
    apt-get install libmysqlclient-dev

Python - python-dev is needed to build the MySQLdb package

    apt-get install python
    apt-get install python-dev

Pip - for installing python packages

    apt-get install python-pip

nfdump - _optional_, for importing Cisco binary NetFlow dumps

    apt-get install nfdump

## Installation

1. Clone the repository
2. Run `pip install -r requirements.txt` from within the directory to install necessary packages.
3. Duplicate `dbconfig.py` as `dbconfig_local.py` and fill out database credentials

## Usage

1. Import log files into database by any combination of the following methods:
   1. Palo Alto logs: `python import_paloalto.py <file>`. The [paloalto syslog](https://www.paloaltonetworks.com/documentation/61/pan-os/pan-os/reports-and-logging/syslog-field-descriptions.html) format is expected.
   2. nfdumps: `python import_nfdump.py <file>` Binary files from **nfcapd** are expected. nfdump must be installed.
   3. Cisco ASA logs: `python import_asa.py <file>`.  Thanks to Emre for contributing. 
   4. AWS VPC Flow logs: `python import_aws.py <file>`. Thanks to Emre for contributing. [VPC log spec](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/flow-logs.html#flow-log-records)  

   Import from all files before going to step 2

2. Run the preprocessing script `python preprocess.py` to analyze and organize the data

3. Start the server locally by running: `python server.py`

4. Navigate your browser to localhost:8080 and use your mouse to explore the network map


changelog.txt
2016-12-19 .. 2016-12-23
Fixed some issues with live mode:
    Links are merged now instead of INSERT/IGNORED
    New LinksIn/Out are recalculated during each preprocess
    removed staging tables for LinksIn/Out
Merged all to master.

2016-12-12 .. 2016-12-16
preprocessing is done in transactions to maintain integrity
log uploading mostly works.
Lots of testing and fixing of bugs.
upgraded live_update.py:
    two threads:
        one for reading the socket
        one for processing in batches
    Fixed issue with inserting duplicates into Nodes table.
    Fills buffers now and processes batches
        instead of fully processing every single line received at once

2016-12-05 .. 2016-12-09
Settings page to switch between data sources
Allows for:
    Selecting data sources,
    Turning on and off auto-refresh on the map
    Creating a new data source
    Deleting an old data source
    Deleting custom data:
        Hostnames
        Tags
        Environments
    Uploading a file-based log to a datasource
Data-source specific tables are now prefixed.
Converted all db-accessing files to use prefixed data sources
MAP: obeys settings now.


2016-11-28 .. 2016-12-02
BACKEND: reformatted node-details requisition endpoint to be more robust
BACKEND: removed "overview" page
QA: updated old test cases
QA: Added filter tests
MAP: sped up details summary query (slightly)
MAP: simplified details query code
METADATA: added protocol support
TABLE: added protocol support
TABLE: fixed styling of filter input
TABLE: download button! csv format data.
MAP: colorized links
MAP: added protocol support
MAP: fixed bug with missing loopback arrowheads

2016-11-21 .. 2016-11-22
Added new data points!
    Bytes, Packets, Dirations, and Protocols (almost)
Merged into Syslog table
Merged into Links Table
Merged into LinksIn/Out Tables
Converted details to list format instead of dictionary.
    Less data to transfer, more generalized
Updated PaloAlto importer to store new information
Updated nfdump importer to store new information
Placeholder updates for ASA and AWS importers
Metadata: updated to display new information
Table View: fixed sorting bug on some columns
Table View: Fixed bug in "apply search" button not working

2016-11-17
added logo images to repo
cleared old metadata when new search begun
fixed bug in disabled filters not working
fixed bug in url list
removed extra debug printing/logging

2106-11-16
sorting tables on metadata page
    (sorting is done in the DB. tablesort.js removed from repository)
fixed several Chrome browser issues
clarified metadata search hint
tableview filters accordian opens automatically if no filters are present

2016-11-15
env filter in tableview drastically faster after adding index to table
updated tests
dynamically generating table headers for details (map selection info and metadata)

2016-11-14
jslinting,
tag displayed in selection list
tags addable, removable
environments added to tableview and metadata
    Can have 1 environment only, but is inherited from higher subnets
added environment filter to tableview

2016-11-10
Added tags table to database
fixed bug with new filters not being added on pressing "apply filters"
added "endpoints" key to metadat quickinfo
reorganized metadata code structure
added hostname editing to metadata page
fixed table formation for "name" row

2016-11-09
Added hostname edit ability to tableview
added role filter to tableview
added tag display support to tableview

2016-11-08
added pagination to metadata
renamed tableview columns for clarity

2016-11-07
completed second-pass quick info for metadata page
added childNodes tab to metadata

2016-11-04
WIP metadata page. Summary information

2016-11-03
added links from the map to the metadata page
added target and mask filters to tableview

2016-11-02
converted to new schema:
    map links
    stats page
    preprocessor
    test cases
Deleted old tables. Syslog no longer gets used by the site after preprocessing.

2016-11-01
Added column sorting on table view
converted details requests to new schema
converted map nodes to new schema
partly converted map links to new schema

2016-10-31
Celebrate Halloween with a new DB schema!
    New tables are Nodes, Links, LinksIn, LinksOut
    Nodes:    ((ipstart, ipend)[PK], subnet, alias, x, y, radius)
    Links:    ((src, dst, port, timestamp)[PK], links)
    LinksIn:  (((src_start, src_end)[FK], (dst_start, dst_end)[FK], port, timestamp)[PK], links)
    LinksOut: (((src_start, src_end)[FK], (dst_start, dst_end)[FK], port, timestamp)[PK], links)
    NOTE: Links no longer include position information--this is calculated in the JS based on Node position
    NOTE: Nodes no longer carry children/connections information. Wasn't being used, wasn't being updated.
    NOTE: LinksIn/Out are basically views, in that they are created from the contents of Links.
        LinksIn/Out are precalculations/preaggregations for mapping purposes.
Added page_size attribute to tables page UI, querystring, python model, and DB query parameters
Added default node display of all nodes when no filter is specified.

2016-10-24 .. 2016-10-28
Frontend:
    Rebuilt the port system
        organized into a class now
        can be used by both map and metadata and future table views.
    Added table view page template
    Added temp filters to UI
    Filters modeled as JS objects
        filter markup is now js-generated
        filters can be created, deleted, summarized
    added "tags" filter
    saved filters in the query string
    encode/decode filters for backend requests
Backend:
    Backend model of filters
    decoded filters from AJAX requests
    developed queries to support filters
    host links from tables page to metadata page
    added pagination of searches
Decided to store all search information in a query string
    This allows the forwards/backwards buttons to work.
    (good) This allows the python backend to build the webpage for the client

filters.py
import common
import dbaccess


class Filter (object):
    def __init__(self, f_type, enabled):
        self.type = f_type
        self.enabled = enabled
        self.params = {}

    def load(self, params):
        keys = self.params.keys()
        keys.sort()
        if len(keys) != len(params):
            print("Wrong number of parameters for constructor")
            raise ValueError("Wrong number of parameters for constructor")
        for index, key in enumerate(keys):
            self.params[key] = params[index]

    def testString(self):
        s = ""
        if self.enabled:
            s += "( 1) "
        else:
            s += "(0 ) "
        s += self.type + "\n\t"
        s += "\n\t".join([": ".join([k, v]) for k, v in self.params.iteritems()])
        return s

    def where(self):
        raise NotImplemented("This method must be overridden in the ({0}) class.".format(self.type))

    def having(self):
        raise NotImplemented("This method must be overridden in the ({0}) class.".format(self.type))


class SubnetFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "subnet", enabled)
        self.params['subnet'] = ""

    def where(self):
        subnet = int(self.params['subnet'])
        valid_subnets = [8, 16, 24, 32]
        if subnet not in valid_subnets:
            raise ValueError("Subnet is not valid. ({net} not in {valid})".format(net=subnet, valid=valid_subnets))

        return "nodes.subnet = {0}".format(subnet)

    def having(self):
        return ""


class MaskFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "mask", enabled)
        self.params['mask'] = ""

    def where(self):
        r = common.determine_range_string(self.params['mask'])
        return "nodes.ipstart BETWEEN {0} AND {1}".format(r[0], r[1])

    def having(self):
        return ""


class PortFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "port", enabled)
        self.params['connection'] = ""
        self.params['port'] = ""
        # connections: 0, 1, 2, 3
        # 0: connects to port n
        # 1: doesn't connect to port n
        # 2: receives connection from port n
        # 3: doesn't receive connection from port n

    def where(self):
        prefix = dbaccess.get_settings_cached()['prefix']

        if self.params['connection'] == '0':
            return "EXISTS (SELECT port FROM {prefix}LinksOut AS `lo` WHERE lo.port = '{0}' && lo.src_start = nodes.ipstart && lo.src_end = nodes.ipend)".format(int(self.params['port']), prefix=prefix)
        elif self.params['connection'] == '1':
            return "NOT EXISTS (SELECT port FROM {prefix}LinksOut AS `lo` WHERE lo.port = '{0}' && lo.src_start = nodes.ipstart && lo.src_end = nodes.ipend)".format(int(self.params['port']), prefix=prefix)

        elif self.params['connection'] == '2':
            return "EXISTS (SELECT port FROM {prefix}LinksIn AS `li` WHERE li.port = '{0}' && li.dst_start = nodes.ipstart && li.dst_end = nodes.ipend)".format(int(self.params['port']), prefix=prefix)
        elif self.params['connection'] == '3':
            return "NOT EXISTS (SELECT port FROM {prefix}LinksIn AS `li` WHERE li.port = '{0}' && li.dst_start = nodes.ipstart && li.dst_end = nodes.ipend)".format(int(self.params['port']), prefix=prefix)
        else:
            print ("Warning: no match for connection parameter of PortFilter when building WHERE clause. "
                   "({0}, type: {1})".format(self.params['connection'], type(self.params['connection'])))
        return ""

    def having(self):
        return ""


class ConnectionsFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "connections", enabled)
        self.params['comparator'] = ""
        self.params['direction'] = ""
        self.params['limit'] = ""

    def where(self):
        return ""

    def having(self):
        HAVING = ""
        if self.params['comparator'] not in ['=', '<', '>']:
            return ""
        limit = int(self.params['limit'])
        if self.params['direction'] == "i":
            src = "conn_in"
        else:
            src = "conn_out"

        HAVING += "{src} {comparator} '{limit}'".format(src=src, comparator=self.params['comparator'], limit=limit)
        return HAVING


class TargetFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "target", enabled)
        self.params['target'] = ""
        self.params['to'] = ""
        # for variable `to`:
        #   0: hosts connect to target
        #   1: hosts do NOT connect to target
        #   2: hosts receive connections from target
        #   3: hosts do NOT receive connections from target

    def where(self):
        ipstart, ipend = common.determine_range_string(self.params['target'])
        prefix = dbaccess.get_settings_cached()['prefix']
        if self.params['to'] == '0':
            return "EXISTS (SELECT 1 FROM {prefix}Links AS `l` WHERE l.dst BETWEEN {lower} AND {upper} " \
                   "AND l.src BETWEEN nodes.ipstart AND nodes.ipend)".format(lower=ipstart, upper=ipend, prefix=prefix)
        elif self.params['to'] == '1':
            return "NOT EXISTS (SELECT 1 FROM {prefix}Links AS `l` WHERE l.dst BETWEEN {lower} AND {upper} " \
                   "AND l.src BETWEEN nodes.ipstart AND nodes.ipend)".format(lower=ipstart, upper=ipend, prefix=prefix)
        
        if self.params['to'] == '2':
            return "EXISTS (SELECT 1 FROM {prefix}Links AS `l` WHERE l.src BETWEEN {lower} AND {upper} " \
                   "AND l.dst BETWEEN nodes.ipstart AND nodes.ipend)".format(lower=ipstart, upper=ipend, prefix=prefix)
        elif self.params['to'] == '3':
            return "NOT EXISTS (SELECT 1 FROM {prefix}Links AS `l` WHERE l.src BETWEEN {lower} AND {upper} " \
                   "AND l.dst BETWEEN nodes.ipstart AND nodes.ipend)".format(lower=ipstart, upper=ipend, prefix=prefix)
        else:
            print ("Warning: no match for 'to' parameter of TargetFilter when building WHERE clause. "
                   "({0}, type: {1})".format(self.params['to'], type(self.params['to'])))
        return ""

    def having(self):
        return ""


class TagsFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "tags", enabled)
        self.params['has'] = ""
        self.params['tags'] = ""

    def where(self):
        tags = str(self.params['tags']).split(",")
        phrase = "EXISTS (SELECT 1 FROM Tags WHERE tag={0} AND Tags.ipstart <= nodes.ipstart AND Tags.ipend >= nodes.ipend)"
        if self.params['has'] != '1':
            phrase = "NOT " + phrase
        return "\n AND ".join([phrase.format(common.web.sqlquote(tag)) for tag in tags])

    def having(self):
        return ""


class EnvFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "env", enabled)
        self.params['env'] = ""

    def where(self):
        return ""

    def having(self):
        env = self.params['env']
        return "env = {0}".format(common.web.sqlquote(env))


class RoleFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "role", enabled)
        self.params['comparator'] = ""
        self.params['ratio'] = ""

    def where(self):
        return ""

    def having(self):
        cmp = self.params['comparator']
        if cmp not in ['<', '>']:
            cmp = '<'
        ratio = float(self.params['ratio'])
        return "(conn_in / (conn_in + conn_out)) {0} {1:.4f}".format(cmp, ratio)


class ProtocolFilter(Filter):
    def __init__(self, enabled):
        Filter.__init__(self, "protocol", enabled)
        self.params['handles'] = ""
        self.params['protocol'] = ""

    def where(self):
        return ""

    def having(self):
        handles = '0'
        if self.params['handles'] in ['0', '1', '2', '3']:
            handles = self.params['handles']
        protocol = common.web.sqlquote("%" + self.params['protocol'] + "%")

        if handles == '0':
            return "proto_in LIKE {protocol}".format(protocol=protocol)
        if handles == '1':
            return "proto_in NOT LIKE {protocol}".format(protocol=protocol)
        if handles == '2':
            return "proto_out LIKE {protocol}".format(protocol=protocol)
        if handles == '3':
            return "proto_out NOT LIKE {protocol}".format(protocol=protocol)

        return ""


filterTypes = [SubnetFilter,PortFilter,ConnectionsFilter,TagsFilter,MaskFilter,TargetFilter,RoleFilter,EnvFilter,ProtocolFilter]
filterTypes.sort(key=lambda x: str(x)) #sort classes by name


def readEncoded(filterString):
    filters = []
    for encodedFilter in filterString.split("|"):
        try:
            params = encodedFilter.split(";")
            typeIndex, enabled, params = params[0], params[1], params[2:]
            enabled = (enabled == "1") #convert to boolean
            filterClass = filterTypes[int(typeIndex)]
            f = filterClass(enabled)
            f.load(params)
            filters.append(f)
        except:
            print("ERROR: unable to decode filter: " + encodedFilter)
    return filters